### PR TITLE
(PUP-7620) Add the unique function

### PR DESCRIPTION
**Pull request by hlindberg**

### CODE REVIEW

File: lib/puppet/functions/unique.rb [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+# ~~~
+#
+# @example Using unique with an Array
+#
+# ~~~puppet
+# # will produce ['a', 'b', 'c']
+# ['a', 'b', 'c', 'a', 'a', 'b'].unique
+# ~~~
+#
+# @example Using unique with a Hash
```
Comment (thallgren) [inline lib/puppet/functions/unique.rb]: When is this useful?

File: lib/puppet/functions/unique.rb [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+# * If the argument is a `String`, the unique set of characters are returned as a new `String`.
+# * If the argument is a `Hash`, the resulting hash associates a set of keys with a set of unique values.
+# * For all other types of `Iterable` (`Array`, `Iterator`) the result is an `Array` with
+#   a unique set of entries.
+# * Comparison of all `String` values are case sensitive.
+# * An optional code block can be given - if present it is given each candidate value and its return is used instead of the given value. This
+#   enables transformation of the value before comparison. The result of the lambda is only used for comparison.
+# * The optional code block when used with a hash is given each value (not the keys).
+#
+# @example Using unique with a String
```
Comment (thallgren) [inline lib/puppet/functions/unique.rb]: When is this useful?

**Comment by thallgren** [comment]
Travis fails with some Ruby versions.

File: lib/puppet/functions/unique.rb (hunk 2 above)
Comment (hlindberg) [inline lib/puppet/functions/unique.rb]: It is for backwards compatibility reasons (the stdlib function has this), and I imagine it to be useful for handling something like a set of optional flags to a command (to ensure the same flag is not used more than once).

File: lib/puppet/functions/unique.rb (hunk 1 above)
Comment (hlindberg) [inline lib/puppet/functions/unique.rb]: When you are interested in the set of keys for which values are unique. `{ john => brown, mary => mauve, fred => mauve}` produces `{[john] => [brown], [mary, fred] => [mauve]}´

**Pull request merged**